src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,405 @@
+    private HkdfKeyDerivation(String hmacAlgName, int hmacLen,
+                      KDFParameters kdfParameters)
+        throws InvalidAlgorithmParameterException {
+        super(kdfParameters);
+        if (kdfParameters != null) {
+            throw new InvalidAlgorithmParameterException(
+                hmacAlgName + " does not support parameters");
+        }
+        // added to enforce valid values at reviewer's request
+        if (!Arrays.asList(SUPPORTED_HMAC_SIZES).contains(hmacLen)){
+            throw new InternalError(
+                "Subclass attempted to use an invalid hmacLen");
+        }
+        this.hmacAlgName = hmacAlgName;
+        this.hmacLen = hmacLen;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     *
+     * @return a derived {@code SecretKey} object of the specified algorithm
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationParameterSpec} is
+     *     invalid or if the combination of {@code alg} and the {@code derivationParameterSpec}
+     *     results in something invalid, ie - a key of inappropriate length
+     *     for the specified algorithm
+     * @throws NoSuchAlgorithmException
+     *     if {@code alg} is empty or invalid
+     * @throws IllegalArgumentException
+     *     if {@code alg} is {@code null} or empty
+     */
+    @Override
+    protected SecretKey engineDeriveKey(String alg,
+                                        AlgorithmParameterSpec derivationSpec)
+        throws InvalidAlgorithmParameterException, NoSuchAlgorithmException {
+
+        if (alg == null) {
+            throw new NullPointerException(
+                "the algorithm for the SecretKey return value must not be null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                "the algorithm for the SecretKey return value must not be "
+                + "empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(derivationSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived keying material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationSpec)
+        throws InvalidAlgorithmParameterException {
+        List<SecretKey> ikms, salts;
+        byte[] inputKeyMaterial, salt, pseudoRandomKey, info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
```

Comment:
This commentary about backporting doesn't belong in the code comments. While it is desirable to avoid dependencies in _APIs_ for features that might be backported, it is acceptable to use new JDK features in the _implementation_, especially if they make the code more efficient, readable, etc.